### PR TITLE
jqls: Adds jq language server

### PR DESCRIPTION
**Pull request by dgmcdona**
This PR adds support for the [jq language server](https://github.com/wader/jq-lsp) by @wader.

### CODE REVIEW

**Comment by dgmcdona** [comment]
Just saw your comment on another PR about unofficial language servers - I understand if you don't want to add this one for that reason. It is currently the only language server available for jq, FWIW.

**Comment by wader** [comment]
Yeap understandable. Can add that sadly the state official jq (github.com/stedolan/jq) has stalled for a long time. There are efforts to continue the development in an official manner but it's waiting for stedolan to approve as i understand it.

File: lua/mason-registry/jqls/init.lua [hunk 1]
```diff
@@ -0,0 +1,15 @@
+local Pkg = require "mason-core.package"
+local _ = require "mason-core.functional"
+local go = require "mason-core.managers.go"
+
+return Pkg.new {
+    name = "jqls",
+    desc = _.dedent [[
+        jqls is a language server for the jq language, developed by Matthias Wader. It provides
```
Comment (wader) [inline lua/mason-registry/jqls/init.lua:8]: "Matthias Wade" is a cool name but sadly my real name is "Mattias Wadman" :)

Comment (dgmcdona) [reply lua/mason-registry/jqls/init.lua:8]: Sorry!!

Comment (wader) [reply lua/mason-registry/jqls/init.lua:8]: No worries :)

Comment (dgmcdona) [reply lua/mason-registry/jqls/init.lua:8]: It was the mwader twitter handle that threw me off

Comment (wader) [reply lua/mason-registry/jqls/init.lua]: Aha yeah someone already had @wader :(

**Comment by williamboman** [comment]
@williambotman /fixup

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/lua/mason-registry/jqls/init.lua b/lua/mason-registry/jqls/init.lua
deleted file mode 100644
index 60f6dc6..0000000
--- a/lua/mason-registry/jqls/init.lua
+++ /dev/null
@@ -1,15 +0,0 @@
-local Pkg = require "mason-core.package"
-local _ = require "mason-core.functional"
-local go = require "mason-core.managers.go"
-
-return Pkg.new {
-    name = "jqls",
-    desc = _.dedent [[
-        jqls is a language server for the jq language, developed by Mattias Wadman. It provides
-        IDE features to any LSP-compatible editor.
-    ]],
-    homepage = "https://github.com/wader/jq-lsp",
-    languages = { Pkg.Lang.Jq },
-    categories = { Pkg.Cat.LSP },
-    install = go.packages { "github.com/wader/jq-lsp", bin = { "jq-lsp" } },
-}
```

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/lua/mason-registry/jq-lsp/init.lua b/lua/mason-registry/jq-lsp/init.lua
new file mode 100644
index 0000000..f96ec79
--- /dev/null
+++ b/lua/mason-registry/jq-lsp/init.lua
@@ -0,0 +1,15 @@
+local Pkg = require "mason-core.package"
+local _ = require "mason-core.functional"
+local go = require "mason-core.managers.go"
+
+return Pkg.new {
+    name = "jq-lsp",
+    desc = _.dedent [[
+        jq-lsp is a language server for the jq language, developed by Mattias Wadman. It provides IDE features to any
+        LSP-compatible editor.
+    ]],
+    homepage = "https://github.com/wader/jq-lsp",
+    languages = { Pkg.Lang.Jq },
+    categories = { Pkg.Cat.LSP },
+    install = go.packages { "github.com/wader/jq-lsp", bin = { "jq-lsp" } },
+}
```

**Review by williamboman — APPROVED** [review]
Cool! I clumsily renamed the package to `jq-lsp` via @williambotman to better match the upstream name (mason.nvim doesn't track lspconfig server names). Let me know if you think this was an error and I'll revert, otherwise LGTM!

**Comment by williamboman** [comment]
@williambotman /fixup

**Comment by williamboman** [comment]
>Just saw your comment on another PR about unofficial language servers - I understand if you don't want to add this one for that reason. It is currently the only language server available for jq, FWIW.

This had to do with the source of the installation itself. The comment was just raising concerns about repackaging the official source as a seemingly unaffiliated 3rd party npm package.

**Comment by dgmcdona** [comment]
Had named it `jqls` to be consistent with this recent PR in nvim-lspconfig: https://github.com/neovim/nvim-lspconfig/pull/2333

**Review by xiantang — APPROVED** [review]

**Comment by wader** [comment]
Thanks everyone for working on this! What would be appropriate instructions to add to jq-lsp how to use this?

Im till a bit confused about the relation between mason and nvim-lspconfig, is the change to language.lua the nvim-lspconfig config to use? also if one wants jq syntax highlighting can mason be used to install that also?

**Comment by williamboman** [comment]
> Thanks everyone for working on this! What would be appropriate instructions to add to jq-lsp how to use this?
> 
> Im till a bit confused about the relation between mason and nvim-lspconfig, is the change to language.lua the nvim-lspconfig config to use? also if one wants jq syntax highlighting can mason be used to install that also?

Mason is a fully independent and stand-alone plugin for managing external tooling. It has no relation with nvim-lspconfig at all. nvim-lspconfig is primarily a bundle of community-managed configurations for language servers. It also provides mechanisms for starting/attaching to language servers (this also now exists in neovim core `:h vim.lsp.start()`).

As for syntax highlighting there are generally two options. First one is to write a vim syntax file that matches regular expressions to highlight groups (`:h syn-files`). The other option is to leverage tree-sitter. For this you'll first need a parser (see [this one for JSON for example](https://github.com/tree-sitter/tree-sitter-json/blob/master/grammar.js)) and then you'll need to [provide queries that allows Neovim's tree-sitter client to understand the AST](https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/json), for example [highlighting](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/json/highlights.scm).

Tree-sitter is the default mechanism for highlights in Neovim (and it enables other richer editor features via its AST) as of some time ago.

**Comment by wader** [comment]
@williamboman thanks for the explaining, the tree-sitter queries fill in a gap for me, was a bit confused how it fit together

**Pull request merged**